### PR TITLE
Don't require program_id to passed down cpi call chain

### PR DESCRIPTION
**Pull request by seanyoung**
#### Problem

When using cpi, the callee program id must be passed down the call chain.  Remove this restriction. See #34057

#### Summary of Changes

Find the program_id in the transaction context rather than in the instruction context.

Fixes #

### CODE REVIEW

File: programs/sbf/tests/programs.rs [hunk 1]
```diff
@@ -1232,6 +1235,69 @@ fn test_program_sbf_caller_has_access_to_cpi_program() {
...
+        &mut bank_client,
+        &bpf_loader::id(),
+        &mint_keypair,
+        "solana_sbf_rust_caller_access",
+    );
+
+    // Here we are not passing caller2_pubkey down, but cpi should still succeed
+    let account_metas = vec![
+        AccountMeta::new_readonly(caller_pubkey, false),
+        AccountMeta::new_readonly(caller2_pubkey, false),
```
Comment (Lichtso) [inline programs/sbf/tests/programs.rs:1258]: But this is passing `caller2_pubkey` down right here.

**Review by Lichtso — COMMENTED** [review]
Some nits, otherwise looking good.

File: programs/sbf/tests/programs.rs (hunk 1 above)
Comment (seanyoung) [reply programs/sbf/tests/programs.rs:1260]: Yes, that was not great. I've created some better tests below

File: sdk/src/feature_set.rs [hunk 2]
```diff
@@ -910,6 +914,7 @@ lazy_static! {
         (validate_fee_collector_account::id(), "validate fee collector account #33888"),
         (disable_rent_fees_collection::id(), "Disable rent fees collection #33945"),
         (enable_zk_transfer_with_fee::id(), "enable Zk Token proof program transfer with fee"),
+        (program_id_not_required_in_instruction::id(), "Don't require a program id to be passed down the cpi chain"),
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply sdk/src/feature_set.rs]: I've struggled to  come up with a good name. I've called it `no_need_for_program_key_in_parent_instruction`. I'm open to ideas here, it's not great

File: programs/sbf/tests/programs.rs [hunk 3]
```diff
@@ -1232,6 +1235,69 @@ fn test_program_sbf_caller_has_access_to_cpi_program() {
...
+    ];
+    let instruction = Instruction::new_with_bytes(caller_pubkey, &[1], account_metas);
+    let result = bank_client.send_and_confirm_instruction(&mint_keypair, instruction);
+    assert!(result.is_ok(), "{result:?}");
+
+    // Trying cpi into account that does not exist
+    let random_pubkey = Pubkey::from_str("FmuzDmVbvBB7kccuJfdbD6ADU5zsN33nVYb1Y9QDUh77").unwrap();
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply programs/sbf/tests/programs.rs]: Good point, fixed

**Comment by willhickey** [comment]
This repository is no longer in use. Please re-open this pull request in the agave repo: https://github.com/anza-xyz/agave

**Pull request closed without merging**